File: dynamice programming/1143.py
# Microsoft, Amazon, paytm, factset, citrix
# DP on string
# LCS = longest common sub-sequence

# Top-Down(Recursion+Memoization)
# Tc = O(2^m * 2^n) => O(2^m+n) after memoization O((m+1)*(n+1)) => O(m*n)
# Sc = O(m*n)
class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.t = [[-1] * 1001 for _ in range(1001)]

    def solve(self, s1, s2, i, j):
        if i >= self.m or j >= self.n:
            return 0
        if self.t[i][j] != -1:
            return self.t[i][j]
        if s1[i] == s2[j]:
            self.t[i][j] = 1 + self.solve(s1, s2, i + 1, j + 1)
        else:
            self.t[i][j] = max(self.solve(s1, s2, i, j + 1), self.solve(s1, s2, i + 1, j))
        return self.t[i][j]

    def longestCommonSubsequence(self, s1, s2):
        self.m = len(s1)
        self.n = len(s2)
        return self.solve(s1, s2, 0, 0)
    

File: dynamice programming/1143.js
// Top-down
class Solution {
    constructor() {
        this.m = 0;
        this.n = 0;
        this.t = Array.from({ length: 1001 }, () => Array(1001).fill(-1));
    }

    solve(s1, s2, i, j) {
        if (i >= this.m || j >= this.n) {
            return 0;
        }
        if (this.t[i][j] !== -1) {
            return this.t[i][j];
        }
        if (s1[i] === s2[j]) {
            this.t[i][j] = 1 + this.solve(s1, s2, i + 1, j + 1);
        } else {
            this.t[i][j] = Math.max(this.solve(s1, s2, i, j + 1), this.solve(s1, s2, i + 1, j));
        }
        return this.t[i][j];
    }

    longestCommonSubsequence(s1, s2) {
        this.m = s1.length;
        this.n = s2.length;
        return this.solve(s1, s2, 0, 0);
    }
}